velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -57,6 +91,111 @@ void ensureRegexIsCompatible(
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
+/// optional paramter position is provided, only make replacements
+/// after that positon in the string (1 indexed).
+///
+/// If position <= 0, throw error.
+/// If position > length string, return string.
+template <typename T>
+struct RegexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex_(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    result.resize(string.size());
+    std::memcpy(result.data(), string.data(), string.size());
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK_LT(-1, position - 1);
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex_(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece(stringInput.data(), stringInput.size());
+
+    const char* prefixEnd = inputStringPiece.data() +
+        std::min(stringInput.size(), static_cast<size_t>(position - 1));
+    re2::StringPiece prefix(
+        inputStringPiece.data(), prefixEnd - inputStringPiece.data());
+
+    re2::StringPiece targetStringPiece(
+        prefixEnd, inputStringPiece.end() - prefixEnd);
+
+    std::string targetString(
+        targetStringPiece.data(), targetStringPiece.size());
+    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    result.resize(prefix.size() + targetString.size());
+    std::memcpy(result.data(), prefix.data(), prefix.size());
+    std::memcpy(
+        result.data() + prefix.size(),
+        targetString.data(),
+        targetString.size());
+  }
+
+ private:
+  mutable std::unordered_map<std::string, std::unique_ptr<re2::RE2>>
+      patternCache_;
+
+  re2::RE2* getCachedRegex_(const std::string& pattern) const {
+    auto it = patternCache_.find(pattern);
+    if (it != patternCache_.end()) {
+      return it->second.get();
+    }
+
+    auto patternRegex = std::make_unique<re2::RE2>(pattern);
+    auto* rawPatternRegex = patternRegex.get();
+    patternCache_[pattern] = std::move(patternRegex);
```

Comment:
Let's make sure there is a hard cap on the total number of regexes.

Reply:
@mbasmanova Okay, I can do that. Is there a number you have in mind? I can leverage the size of the cache to check, and then throw an exception if it exceeds the limit. Off the top of my head, I can't think of a way to throw that exception while also finishing any processing for future repeated regexes, but I am not sure that matters.

Reply:
> Is there a number you have in mind? 

I don't have a good reference, but maybe we can start with 20.

Reply:
Okay, I implemented a kMaxCompiledRegexes in velox/functions/lib/Re2Functions.h so any regex function can leverage it at a later date. I also added two test cases. One that checks if a VeloxRuntimeError is thrown when we reach the limit. This one runs successfully. And I added another that repeats that use case with slight changes. 

Instead of passing the limit, it hits the limit and then concatenates the RowVector 50,000 times, so we do a total of 20 * 50000 = 1,000,000 replacements but never reach the limit. This runs in 2197ms on my machine and exists successfully.